src/query_plan/query_planning_traversal.rs

```diff
@@ -86,3 +88,116 @@ struct BestQueryPlanInfo {
+                let edge = child.edge.expect("The root edge should not be None");
+                let (_source_node, target_node) = path_tree.graph.edge_endpoints(edge)?;
+                let target_node = path_tree.graph.node_weight(target_node)?;
+                let subgraph_name = &target_node.source;
+                // The edge tail type is one of the subgraph root type, so it has to be an ObjectType.
+                //     const rootType = edge.tail.type as ObjectType;
+                let root_type = match &target_node.type_ {
+                    QueryGraphNodeType::SchemaType(position) => match position {
+                        OutputTypeDefinitionPosition::Object(object) => object.clone().into(),
+                        _ => todo!(), // Return a FederationError?
+                    },
+                    QueryGraphNodeType::FederatedRootType(_) => todo!(), // Return a FederationError?
+                };
+                let group = dependency_graph.get_or_create_root_fetch_group(
+                    subgraph_name,
+                    self.root_kind,
+                    root_type,
+                );
+                compute_groups_for_tree(
+                    dependency_graph,
+                    &child.tree,
+                    // group,
+                    // GroupPath.empty(),
+                    // emptyDeferContext,
+                );
+            }
+        } else {
+            let query_graph_node = path_tree.graph.node_weight(path_tree.node)?;
+            let subgraph_name = &query_graph_node.source;
+            // The edge tail type is one of the subgraph root type, so it has to be an ObjectType.
+            let root_type = match &query_graph_node.type_ {
+                QueryGraphNodeType::SchemaType(position) => {
+                    position.clone().try_into().map_err(|err| {
+                        // TODO: add details to the error message?
+                        // format!("Should not have condition on non-selectable type {position}")
+                        err
+                    })?
+                }
+                QueryGraphNodeType::FederatedRootType(_) => todo!(), // Return a FederationError?
+            };
+            let group = dependency_graph.get_or_create_root_fetch_group(
+                subgraph_name,
+                self.root_kind,
+                root_type,
+            );
+            compute_groups_for_tree(
+                dependency_graph,
+                path_tree,
+                // group,
+                // GroupPath.empty(),
+                // emptyDeferContext,
+            );
+        }
+        Ok(())
+    }
+}
+
+fn compute_groups_for_tree(
+    dependency_graph: &mut FetchDependencyGraph,
+    path_tree: &OpPathTree,
+    // start_group: FetchGroup,
+    // initial_group_path: GroupPath,
+    // initial_defer_context: DeferContext,
+    // initial_context: PathContext = emptyContext,
+)
+//   -> Vec<FetchGroup>
```

Comment:
Do you have a suggestion for Rust data structures here?

`FetchGroup` in particular is central to TypeScript `FetchDependencyGraph` but does not appear to have an equivalent in Rust. A `FetchGroup` represents a node in the graph together with its outgoing edges, right? But Rust relies on petgraph instead

Reply:
We've decided to rename `FetchGroup` in the TypeScript code to `FetchDependencyGraphNode` in the Rust code, since the name `FetchGroup` never really reflected that they were just nodes in a (graph-theory) graph and has led to some confusion (if you look at `fetch_dependency_graph.rs`, there's a `PORT_NOTE` comment describing the rename on `FetchDependencyGraphNode`; usually when a TypeScript function/data structure has its name changed, I try to leave a `PORT_NOTE` mentioning it).